keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,266 @@
+  Request []byte
+
+  // usages specifies a set of usage contexts the key will be
+  // valid for.
+  // See: https://tools.ietf.org/html/rfc5280#section-4.2.1.3
+  //      https://tools.ietf.org/html/rfc5280#section-4.2.1.12
+  Usages []KeyUsage
+
+  // Information about the requesting user.
+  // See user.Info interface for details.
+  Username string
+  // UID information about the requesting user.
+  // See user.Info interface for details.
+  UID string
+  // Group information about the requesting user.
+  // See user.Info interface for details.
+  Groups []string
+  // Extra information about the requesting user.
+  // See user.Info interface for details.
+  Extra map[string]string
+}
+
+// ExtraValue masks the value so protobuf can generate
+type ExtraValue []string
+
+type CertificateSigningRequestStatus struct {
+  // Conditions applied to the request, such as approval or denial.
+  Conditions []CertificateSigningRequestCondition
+
+  // If request was approved, the controller will place the issued
+  // certificate here.
+  Certificate []byte
+}
+
+type CertificateSigningRequestCondition struct {
+  // request approval state, currently Approved or Denied.
+  Type RequestConditionType
+  // brief reason for the request state
+  Reason string
+  // human readable message with details about the request state
+  Message string
+}
+
+type RequestConditionType string
+
+// These are the possible conditions for a certificate request.
+const (
+  CertificateApproved RequestConditionType = "Approved"
+  CertificateDenied   RequestConditionType = "Denied"
+)
+```
+
+### Manual CSR Approval With Kubectl
+
+A Kubernetes administrator (with appropriate permissions) can manually approve
+(or deny) Certificate Signing Requests by using the `kubectl certificate
+approve` and `kubectl certificate deny` commands.
+
+### Automatic CSR Approval Implementations
+
+The kube-controller-manager ships with an in-built
+[approver](https://github.com/kubernetes/kubernetes/blob/32ec6c212ec9415f604ffc1f4c1f29b782968ff1/pkg/controller/certificates/approver/sarapprove.go)
+for Kubelet TLS Bootstrap that delegates various permissions on CSRs for node
+credentials to authorization. It does this by posting subject access reviews to
+the API server. It punts on TLS certificates for server authentication of the
+Kubelet API because verifying IP SANs for Kubelets in a generic way poses
+challenges.
+
+The GCP controller manager replaces the in-built approver with an
+[approver](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_approver.go)
+that handles TLS certificates for server authentication and support for
+verification of CSRs with proof of control of GCE vTPM.
+
+An external project, [kapprover](https://github.com/coreos/kapprover), does
+policy based approval of kubelet CSRs.
+
+<!--
+TODO: various things use the certificates API for stuff other than Kubelet TLS
```

Comment:
suggest dropping this from the initial backfill merge, and adding it in a later PR